File: python-algorithm/ds_1_stack_queue/L71-m.cpp
//
// Created by 李恒 on 2020/7/2.
//
#include <iostream>
#include <stack>
#include <cstring>

using namespace std;

class Solution {
public:
    string simplifyPath(string path) {
        stack<string> stk;
        string res = "";
        string tem = "";
        if(path[path.size()-1] !='/')
            path += '/';

        for(int i=0;i<path.size();i++){
            if(path[i] != '/'){
                tem += path[i];
            }
            else{
                if(tem != ""){
                    if(tem == "."){}
                    else if(tem == ".."){
                        if(!stk.empty())
                            stk.pop();
                    }
                    else
                        stk.push(tem);
                }
                tem = "";
            }
        }
        if(stk.empty())
            return "/";
        while(!stk.empty()){
            res = '/' + stk.top() + res;
            stk.pop();
        }
        return res;
    }
};

File: python-algorithm/ds_1_stack_queue/L71-m.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/2 8:09
# @Author  : Heng Li
# @File    : L71-m.py
# @Software: PyCharm

# 思路就是把所有'/'之间的目录压入栈，如果是'.' '..' 特殊处理
class Solution:
    def simplifyPath(self, path: str) -> str:
        dir_ls = [p for p in path.split('/') if p != '']
        stk = []
        res = ""
        for i in range(len(dir_ls)):
            if dir_ls[i] == '.':
                continue
            elif dir_ls[i] == '..' :
                if stk != []:
                    stk.pop(-1)
            else:
                stk.append(dir_ls[i])
        if stk == []:
            return "/"
        else:
            # 因为用list模拟stack所以不使用Python的join()
            while stk:
                res = '/'+ stk.pop(-1) + res
            return res